File: dh/tests/draft/io_conc/sample.go
package main

import (
	"fmt"
	"sync"
	"time"
)

var printMu sync.Mutex

func report(label string, format string, args ...any) {
	printMu.Lock()
	defer printMu.Unlock()
	fmt.Printf("[%s] "+format+"\n", append([]any{label}, args...)...)
}

func count(n int, interval time.Duration, label string, started chan<- struct{}) float64 {
	start := time.Now()
	report(label, "before loop %.1f", interval.Seconds())
	close(started)

	for i := 0; i < n; i++ {
		time.Sleep(interval)
		report(label, "slept %.1f | i: %d < n: %d", interval.Seconds(), i, n)
	}

	elapsed := time.Since(start).Seconds()
	report(label, "after loop %.1f", elapsed)
	return elapsed
}

func main() {
	fmt.Println("begin")

	taskA := make(chan float64, 1)
	taskB := make(chan float64, 1)
	startedA := make(chan struct{})
	startedB := make(chan struct{})

	go func() { taskA <- count(2, time.Second, "task a", startedA) }()
	<-startedA
	go func() { taskB <- count(3, 600*time.Millisecond, "task b", startedB) }()
	<-startedB

	total := <-taskA + <-taskB

	fmt.Println("end")
	fmt.Printf("total: %.1f\n", total)
}

File: dh/tests/draft/io_conc/sample.zig
const std = @import("std");

const Io = std.Io;

const Sys = struct {
    io: Io,
};

fn report(label: []const u8, comptime fmt: []const u8, args: anytype) void {
    std.debug.print("[{s}] " ++ fmt ++ "\n", .{label} ++ args);
}

fn count(sys: Sys, n: usize, interval: Io.Duration, label: []const u8) f64 {
    const start = Io.Timestamp.now(sys.io, .awake);
    report(label, "before loop {d:.1}", .{secs(interval)});

    for (0..n) |i| {
        Io.sleep(sys.io, interval, .awake) catch {};
        report(label, "slept {d:.1} | i: {d} < n: {d}", .{ secs(interval), i, n });
    }

    const elapsed = start.durationTo(Io.Timestamp.now(sys.io, .awake));
    report(label, "after loop {d:.1}", .{secs(elapsed)});
    return secs(elapsed);
}

fn secs(duration: Io.Duration) f64 {
    return @as(f64, @floatFromInt(duration.toNanoseconds())) / @as(f64, @floatFromInt(std.time.ns_per_s));
}

pub fn main() !void {
    var debug_allocator: std.heap.DebugAllocator(.{}) = .init;
    defer _ = debug_allocator.deinit();

    var threaded: Io.Threaded = .init(debug_allocator.allocator(), .{});
    defer threaded.deinit();

    const io = threaded.io();
    const sys: Sys = .{ .io = io };

    std.debug.print("begin\n", .{});
    var task_a = Io.async(io, count, .{ sys, 2, Io.Duration.fromSeconds(1), "task a" });
    defer Io.cancel(io, task_a) catch {};
    var task_b = Io.async(io, count, .{ sys, 3, Io.Duration.fromMilliseconds(600), "task b" });
    defer Io.cancel(io, task_b) catch {};

    var total: f64 = 0;
    total += task_a.await(io);
    total += task_b.await(io);

    std.debug.print("end\n", .{});
    std.debug.print("total: {d:.1}\n", .{total});
}
